### PR TITLE
feat: Manage runnable dependencies by HTTP server and improve DB health check

### PR DESCRIPTION
**Pull request by IvayloAndonov**
# Context

Our [current DB health check](https://github.com/input-output-hk/cardano-js-sdk/blob/841cf75d346402028b778e03d2ddb6bcd6a45b1e/packages/cardano-services/src/DbSyncProvider.ts#L4) does not take domain rules into consideration that can infer the health of `cardano-db-sync` :
`const HEALTH_CHECK_QUERY = 'SELECT 1';`

Also currently each provider handles and manages the lifecycle methods of its dependencies, in most cases we use one and the same instance of that dependency (memoized), ex. `OgmiosCardanoNode` dependency now becomes a dependency of all DbSync providers, the same as DB.

Ideally, we would have one single place (`HttpServer`) to instantiate all runnable dependencies and manage them through lifecycle methods.

# Proposed Solution
**Prior Refactoring (ADP-2296)**:
- move `RunnableModule` class out of `cardano-services` package to commonplace (a util package is a good option) and reuse it by multiple packages where is needed
- refactor `OgmiosCardanoNode` to extend `RunnableModule` as a runnable dependency of `HttpServer`
- extend `HttpServerDependencies` with `runnableDependencies` list to enable `HttpServer` to manage all runnable common dependencies across services in one single place through lifecycle methods (initialize, start , shutdown etc.)

**Health check improvement (ADP-2186)**:

- Add CardanoNode as a dependency in DbSyncProvider
- Consider the impact of all providers now being dependent on CardanoNode (Ogmios)
- Compose the health check result based on the following:
1. query `db-sync` for the latest block height (number)
2. fetch `lastKnownTip` via `stateQueryClient` and extract the block no from the node
3. compare the results with a tolerance of 5 blocks behind (`db-sync` could be out of sync slightly)

### CODE REVIEW

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Nice work @IvayloAndonov. Just some minor suggestions

**Review by mkazlauskas — APPROVED** [review]
Nice work!

File: packages/cardano-services/src/Program/utils.ts [hunk 1]
```diff
@@ -29,20 +28,29 @@ export const onFailedAttemptFor =
...
     factor: config.factor,
     maxRetryTime: config.maxRetryTime,
     onFailedAttempt: onFailedAttemptFor(serviceName, logger)
   });
-
 export type DnsResolver = ReturnType<typeof createDnsResolver>;
 
 export const srvRecordToRabbitmqURL = ({ name, port }: SrvRecord) => new URL(`amqp://${name}:${port}`);
+
+export const cardanoNodeDependantServices = new Set([
```
[the remark this answers is not in the document]
Comment (IvayloAndonov) [reply packages/cardano-services/src/Program/utils.ts:46]: Yeah, only the `TxSubmitHttpService` is not currently in that list, as the current implementation, it has two providers [modes](https://github.com/input-output-hk/cardano-js-sdk/blob/daeb139e9caab35707a3282443566b64324ee94d/packages/cardano-services/src/Program/loadHttpServer.ts#L161-L166): without and with the queue.
The first one ([ogmiosTxSubmitProvider](https://github.com/input-output-hk/cardano-js-sdk/blob/9298ee0e6c98bef0c1aa4dc70f349094ec00a5c9/packages/ogmios/src/Provider/TxSubmitProvider/ogmiosTxSubmitProvider.ts#L23-L43)) depends on ogmios but does not depend on `db-sync`.
The second one ([RabbitMqTxSubmitProvider](https://github.com/input-output-hk/cardano-js-sdk/blob/a1d07549e7a5fccd36b9f75b9f713c0def8cb97f/packages/cardano-services/src/TxSubmit/rabbitmq/RabbitMqTxSubmitProvider.ts#L40-L154)) depends only on `rabbitmq`,
that's why this service is the only exception at the moment, we could think if it needs to be extended somehow.

**Review by iccicci — APPROVED** [review]
I see several improvements in this PR.
GJ!

File: packages/cardano-services/src/Http/HttpServer.ts [hunk 2]
```diff
@@ -77,37 +71,36 @@ export class HttpServer extends RunnableModule {
...
     this.app.use((req, _res, next) => {
       const { body, method, path, query } = req;
-
       requestLogger.debug({ body, method, path, query });
       next();
     });
 
+    for (const dependency of this.#dependencies.runnableDependencies) {
+      await dependency.initialize();
+      this.logger.debug(`Runnable dependency ${dependency.name} has been initialized.`);
```
[the remark this answers is not in the document]
Comment (IvayloAndonov) [reply packages/cardano-services/src/Http/HttpServer.ts]: a29e6f3a47c9b7df8d9588ffec712529883f55d0

File: packages/cardano-services/src/Http/HttpServer.ts [hunk 3]
```diff
@@ -137,15 +127,22 @@ export class HttpServer extends RunnableModule {
...
     this.server = await listenPromise(this.app, this.#config.listen);
   }
 
   async shutdownImpl(): Promise<void> {
     this.#healthGauge?.set(0);
-    for (const service of this.#dependencies.services) await service.shutdown();
+
+    for (const service of this.#dependencies.services) {
+      await service.shutdown();
+      this.logger.debug(`Service ${service.name} has been shutdown.`);
```
[the remark this answers is not in the document]
Comment (IvayloAndonov) [reply packages/cardano-services/src/Http/HttpServer.ts]: a29e6f3a47c9b7df8d9588ffec712529883f55d0

File: packages/cardano-services/src/Http/HttpServer.ts [hunk 4]
```diff
@@ -137,15 +127,22 @@ export class HttpServer extends RunnableModule {
...
+    }
+    for (const dependency of this.#dependencies.runnableDependencies) {
+      await dependency.shutdown();
+      this.logger.debug(`Runnable dependency ${dependency.name} has been shutdown.`);
```
[the remark this answers is not in the document]
Comment (IvayloAndonov) [reply packages/cardano-services/src/Http/HttpServer.ts]: a29e6f3a47c9b7df8d9588ffec712529883f55d0

File: packages/cardano-services/src/Program/services/ogmios.ts [hunk 5]
```diff
@@ -135,14 +134,16 @@ export const ogmiosCardanoNodeWithDiscovery = async (
   });
 };
 
-export const getOgmiosCardanoNode = memoize(
```
Comment (IvayloAndonov) [inline packages/cardano-services/src/Program/services/ogmios.ts:138]: Memoize no longer needed

File: packages/cardano-services/test/Program/services/postgres.test.ts [hunk 6]
```diff
@@ -127,12 +123,16 @@ describe('Service dependency abstractions', () => {
         config = { listen: { port } };
         apiUrlBase = `http://localhost:${port}/network-info`;
         epochMonitor = new DbSyncEpochPollService(db!, 1000);
+        lastBlockNoInDb = (await db!.query<BlockNoModel>(findLastBlockNo)).rows[0].block_no;
+        cardanoNode = mockCardanoNode(
+          healthCheckResponseMock({ blockNo: lastBlockNoInDb })
+        ) as unknown as OgmiosCardanoNode;
```
Comment (IvayloAndonov) [inline packages/cardano-services/test/Program/services/postgres.test.ts:129]: This way of mocking will prepare our unit/integration tests when we no longer rely on test fixtures data, but work with local data generated by local-network e2e tests. It handles the introduced relationship between `db-sync` and `cardanoNode` related to the service's health check. (db sync should be synced with the node to determine healthiness)

File: packages/cardano-services/src/Program/utils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rhyslbw) [reply packages/cardano-services/src/Program/utils.ts:46]: Ah yes. I'm thinking about generic HTTP providers, which probably all should have a dependency on the local node, even if it's just for clients to perform consistency checks. It's out of scope for this PR though, so I'm happy with it as-is

**Review by rhyslbw — APPROVED** [review]

**Comment by IvayloAndonov** [comment]
Commits are squashed, CI build passed, and ready for merge once passed again.

**Review by iccicci — APPROVED** [review]

**Review by mkazlauskas — APPROVED** [review]

**Review by rhyslbw — APPROVED** [review]

**Pull request merged**